### PR TITLE
test

### PR DESCRIPTION
**Pull request by mr-lee**

### CODE REVIEW

File: agamemnon/factory.py [hunk 1]
```diff
@@ -464,8 +472,134 @@ def __getattr__(self, item):
         if not item in self.__dict__:
             return getattr(self.delegate, item)
 
+    def search_index(self, type, index_name, query, num_results=0):
+        ns_index_name = str(type) + "__" + index_name
```
Comment (jbryan) [inline agamemnon/factory.py:476]: Let's change the delimiter to something unlikely to be found an a name.  e.g. ## or |

Also, can we make index_name an array so we can query multiple indices?

File: agamemnon/factory.py [hunk 2]
```diff
@@ -464,8 +472,134 @@ def __getattr__(self, item):
...
+        q = WildcardQuery('_all',query)
+        results = self.conn.search(query=q, indices=ns_index_name, doc_types=type)
+        nodelist = []
+        n = 0
+        nodes = sorted(results['hits']['hits'], key=itemgetter('_score'),reverse=True)
+        for r in nodes:
```
Comment (mr-lee) [inline agamemnon/factory.py:482]: write this so that python can preallocate the list

File: agamemnon/factory.py [hunk 3]
```diff
@@ -464,8 +472,134 @@ def __getattr__(self, item):
...
+            key = r['_source']['__id']
+            node = self.get_node(type,key)
+            nodelist.append(node)
+            n+=1
+            if(n>=num_results and num_results != 0):
+                break
+        return nodelist
+
+    def create_index(self, type, args, index_name):
```
Comment (mr-lee) [inline agamemnon/factory.py:491]: args are the attributes indexed over

File: agamemnon/factory.py [hunk 4]
```diff
@@ -464,8 +472,134 @@ def __getattr__(self, item):
...
+        node_list = [rel.target_node for rel in ref_node.instance.outgoing]
+        mapping = self.conn.get_mapping(type,ns_index_name)
+        for node in node_list:
+            key = node.key
+            index_dict = self.populate_index_document(type,ns_index_name,node.attributes,mapping)
+            self.conn.index(index_dict,ns_index_name,type,key)
+        self.conn.refresh([ns_index_name])
+
+    #given a type of node, update all the indices of that type with a new node
+    def update_indices(self,type,attributes,key):
```
Comment (jbryan) [inline agamemnon/factory.py:546]: When moving these methods into a plugin, can we make them take just a node?  This way we can have a uniform plugin api and pass around a single type of object.

File: agamemnon/factory.py [hunk 5]
```diff
@@ -464,8 +472,134 @@ def __getattr__(self, item):
...
+                    'type' : 'nGram',
+                    'max_gram' : 30,
+                    'min_gram' : 1                                 
+                    }                           
+                }
+            }
+        }}
+        self.conn.create_index(ns_index_name,settings)
+        mapping = {}
+        args.append('__id')
```
Comment (mr-lee) [inline agamemnon/factory.py:514]: rewrite to extract the _id field without adding it as an indexed attribute

File: agamemnon/factory.py [hunk 6]
```diff
@@ -464,8 +472,134 @@ def __getattr__(self, item):
...
+        for node in node_list:
+            key = node.key
+            index_dict = self.populate_index_document(type,ns_index_name,node.attributes,mapping)
+            self.conn.index(index_dict,ns_index_name,type,key)
+        self.conn.refresh([ns_index_name])
+
+    #given a type of node, update all the indices of that type with a new node
+    def update_indices(self,type,attributes,key):
+        type_indices = self.get_indices_of_type(type)
+        if(len(type_indices)>0):
```
Comment (mr-lee) [inline agamemnon/factory.py:548]: remove this line

File: agamemnon/factory.py [hunk 7]
```diff
@@ -464,8 +472,134 @@ def __getattr__(self, item):
...
+
+    def populate_index(self, type, index_name, args):
+        #add all the currently existing nodes into the index
+        ns_index_name = str(type) + "__" + index_name
+        ref_node = self.get_reference_node(type)
+        node_list = [rel.target_node for rel in ref_node.instance.outgoing]
+        mapping = self.conn.get_mapping(type,ns_index_name)
+        for node in node_list:
+            key = node.key
+            index_dict = self.populate_index_document(type,ns_index_name,node.attributes,mapping)
```
Comment (mr-lee) [inline agamemnon/factory.py:541]: delete the index document and recreate

File: agamemnon/factory.py (hunk 4 above)
Comment (jbryan) [inline agamemnon/factory.py:546]: Can we change this to "insert_indices" or something similar as it shoould only be called when a node as added

File: agamemnon/factory.py [hunk 8]
```diff
@@ -464,8 +472,134 @@ def __getattr__(self, item):
...
+            self.conn.index(index_dict,ns_index_name,type,key)
```
Comment (jbryan) [inline agamemnon/factory.py:542]: Do we need to delete here?

File: agamemnon/tests/unit_tests.py [hunk 9]
```diff
@@ -21,6 +22,79 @@ def create_node(self, node_type, id):
...
+        self.failUnlessEqual(2,len(nodes_found))
+        #test modify_indices function
+        new_args = ['string','new_attr']
+        self.ds.create_index(node_type,new_args,new_index_name)
+        nodes_found = self.ds.search_index(node_type,new_index_name,'new_value')
+        self.failUnlessEqual(0,len(nodes_found))
+        self.ds.create_node(node_type,'node_2',{'new_attr':'new_value'})
+        new_node2 = self.ds.get_node(node_type,'node_2')
+        nodes_found = self.ds.search_index(node_type,new_index_name,'new_value')
+        self.failUnless(new_node2 in nodes_found)
```
Comment (mr-lee) [inline agamemnon/tests/unit_tests.py:83]: self.failUnless(node_1 not in nodes_found)

**Pull request closed without merging**